src/backend/distributed/utils/resource_lock.c

```diff
@@ -164,6 +166,58 @@ LockShardDistributionMetadata(int64 shardId, LOCKMODE lockMode)
 }
 
 
+/*
+ * LockReferencedReferenceShardDistributionMetadata acquires the given lock
+ * on the reference tables which has a foreign key from the given relation.
+ */
+void
+LockReferencedReferenceShardDistributionMetadata(uint64 shardId, LOCKMODE lock)
+{
+	ListCell *shardCell = NULL;
+	Oid relationId = RelationIdForShard(shardId);
+
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *referencedRelationList = cacheEntry->referencedRelationsViaForeignKey;
+
+	List *shardList = GetSortedReferenceShards(referencedRelationList);
+	foreach(shardCell, shardList)
+	{
+		uint64 shardId = lfirst_oid(shardCell);
+		LockShardDistributionMetadata(shardId, lock);
+	}
+}
+
+
+/*
+ * GetSortedReferenceShards iterates through the given relation list.
+ * Lists the shards of reference tables and returns the list after sorting.
+ */
+static List *
+GetSortedReferenceShards(List *relationList)
+{
+	List *shardList = NIL;
+	ListCell *relationCell = NULL;
+
+	foreach(relationCell, relationList)
+	{
+		Oid relationId = lfirst_oid(relationCell);
+		List *currentShardList = NIL;
+
+		if (PartitionMethod(relationId) != DISTRIBUTE_BY_NONE)
+		{
+			continue;
+		}
+
+		currentShardList = LoadShardList(relationId);
+		shardList = lappend(shardList, linitial(currentShardList));
+	}
+
+	shardList = SortList(shardList, CompareOids);
```

Comment:
`CompareOids` might be tricky. We should probably use `LoadShardIntervalList()` and `CompareShardIntervalsById`.

Reply:
But then don't forget to rename the function as `GetSortedReferenceShardIntervals` and fix the loop `foreach(shardCell, shardList)`.

Or, you could write a new function `CompareShardIds`, and use it here. Pick one approach